File: iosApp/iosApp/ViewModels/ReaderViewModel.swift
import Foundation
import SwiftUI
import SharedLogic

@MainActor
public final class ReaderViewModel: ObservableObject {
    @Published public private(set) var currentChapter: ChapterContent? = nil
    @Published public private(set) var story: Story? = nil
    @Published public private(set) var chapters: [Chapter] = []
    @Published public private(set) var prevChapter: Chapter? = nil
    @Published public private(set) var nextChapter: Chapter? = nil
    @Published public private(set) var wordCount: Int32 = 0
    @Published public private(set) var estimatedReadMinutes: Int32 = 1
    @Published public private(set) var isBookmarked: Bool = false
    @Published public var showTocSheet: Bool = false
    @Published public var fontSize: CGFloat = 18
    @Published public var showControls: Bool = true
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var errorMessage: String? = nil

    private let readerUseCase: ReaderUseCase

    private var currentStoryId: String = ""
    private var currentChapterId: String = ""

    public init(
        readerUseCase: ReaderUseCase = KoinHelper().readerUseCase
    ) {
        self.readerUseCase = readerUseCase
    }

    public func loadChapter(storyId: String, chapterId: String) {
        self.currentStoryId = storyId
        self.currentChapterId = chapterId
        self.isLoading = true
        self.errorMessage = nil

        Task {
            do {
                let parsedStoryId = StoryId.companion.from(compositeValue: storyId)
                let sessionResult = try await readerUseCase.loadChapterSession(
                    storyId: parsedStoryId,
                    chapterId: chapterId,
                    prefetchScope: nil
                )

                if let session = sessionResult.getOrNull() as? ReaderSessionData {
                    self.story = session.story
                    self.currentChapter = session.content
                    self.chapters = session.chapters
                    self.prevChapter = session.prevChapter
                    self.nextChapter = session.nextChapter
                    self.wordCount = session.wordCount
                    self.estimatedReadMinutes = session.estimatedReadMinutes
                    self.errorMessage = nil
                } else if let err = sessionResult.errorOrNull() {
                    self.errorMessage = err.message
                }
                self.isLoading = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }

    public func toggleBookmark() {
        guard let story = story, let currentChapter = chapters.first(where: { $0.id == currentChapterId }) else { return }

        Task {
            do {
                let result = try await readerUseCase.toggleBookmark(story: story, chapter: currentChapter, snippet: nil)
                if let bookmarked = result.getOrNull() as? NSNumber {
                    self.isBookmarked = bookmarked.boolValue
                }
            } catch {
                // Ignore error
            }
        }
    }

    public func toggleTocSheet() {
        showTocSheet.toggle()
    }

    public func increaseFontSize() {
        if fontSize < 32 {
            fontSize += 2
        }
    }

    public func decreaseFontSize() {
        if fontSize > 12 {
            fontSize -= 2
        }
    }

    public func toggleControls() {
        withAnimation(.easeInOut(duration: 0.2)) {
            showControls.toggle()
        }
    }
}

File: androidApp/src/main/kotlin/com/slowbuild/storyverse/storyverse/ui/reader/ReaderViewModel.kt
package com.slowbuild.storyverse.storyverse.ui.reader

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.slowbuild.storyverse.core.result.AppResult
import com.slowbuild.storyverse.domain.model.Chapter
import com.slowbuild.storyverse.domain.model.ChapterContent
import com.slowbuild.storyverse.domain.model.Story
import com.slowbuild.storyverse.domain.model.StoryId
import com.slowbuild.storyverse.domain.usecase.ReaderUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class ReaderUiState(
    val isLoading: Boolean = true,
    val story: Story? = null,
    val currentChapter: Chapter? = null,
    val chapterContent: ChapterContent? = null,
    val chapters: List<Chapter> = emptyList(),
    val prevChapter: Chapter? = null,
    val nextChapter: Chapter? = null,
    val wordCount: Int = 0,
    val estimatedReadMinutes: Int = 1,
    val isBookmarked: Boolean = false,
    val showTocSheet: Boolean = false,
    val fontSizeSp: Float = 17f,
    val errorMessage: String? = null,
    val showControls: Boolean = true
)

class ReaderViewModel(
    private val readerUseCase: ReaderUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(ReaderUiState())
    val uiState: StateFlow<ReaderUiState> = _uiState.asStateFlow()

    private var currentStoryId: StoryId? = null

    fun loadChapter(rawOrCompositeStoryId: String, chapterId: String) {
        val storyId = StoryId.from(rawOrCompositeStoryId)
        currentStoryId = storyId

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }

            when (val sessionResult = readerUseCase.loadChapterSession(storyId, chapterId, viewModelScope)) {
                is AppResult.Success -> {
                    val session = sessionResult.data
                    val bookmarks = readerUseCase.observeBookmarks(storyId).firstOrNull().orEmpty()
                    val isBookmarked = bookmarks.any { it.chapterId == chapterId }

                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            story = session.story,
                            currentChapter = session.currentChapter,
                            chapterContent = session.content,
                            chapters = session.chapters,
                            prevChapter = session.prevChapter,
                            nextChapter = session.nextChapter,
                            wordCount = session.wordCount,
                            estimatedReadMinutes = session.estimatedReadMinutes,
                            isBookmarked = isBookmarked,
                            errorMessage = null
                        )
                    }
                }
                is AppResult.Error -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = sessionResult.error.message
                        )
                    }
                }
            }
        }
    }

    fun toggleBookmark() {
        val story = _uiState.value.story ?: return
        val chapter = _uiState.value.currentChapter ?: return

        viewModelScope.launch {
            when (val res = readerUseCase.toggleBookmark(story, chapter)) {
                is AppResult.Success -> {
                    _uiState.update { it.copy(isBookmarked = res.data) }
                }
                is AppResult.Error -> Unit
            }
        }
    }

    fun toggleTocSheet() {
        _uiState.update { it.copy(showTocSheet = !it.showTocSheet) }
    }

    fun hideTocSheet() {
        _uiState.update { it.copy(showTocSheet = false) }
    }

    fun toggleControls() {
        _uiState.update { it.copy(showControls = !it.showControls) }
    }

    fun increaseFontSize() {
        _uiState.update { it.copy(fontSizeSp = (it.fontSizeSp + 1.5f).coerceAtMost(28f)) }
    }

    fun decreaseFontSize() {
        _uiState.update { it.copy(fontSizeSp = (it.fontSizeSp - 1.5f).coerceAtLeast(13f)) }
    }
}
